rfcs/convergence/slot-null-rendering.md

```diff
@@ -0,0 +1,261 @@
+      icon: 'i',
+    },
+    button: resolveShorthand(props.button, {
+      children: React.Fragment, // 🚨 getSlots requires children
+    }),
+    icon: resolveShorthand(props.icon),
+  };
+
+  const { slots, slotProps } = getSlots(state, ['input']);
+
+  return (
+    <slots.root {...slotProps.root}>
+      <slots.button {...slotProps.button}>
+        <slots.icon {...slotProps.icon} />
+      </slots.button>
+    </slots.root>
+  );
+}
+```
+
+[Live Example](https://stackblitz.com/edit/react-ts-wwguyp?file=SlotAsParent.tsx)
+
+### Native elements without children
+
+e.g: An Icon that doesn't have children `<i class="icon"/>` (this cases is hypothetical, but quite possible, as we cannot verify it in converged components)
+
+## Detailed Design or Proposal
+
+Both options here proposed by the end are equivalent, the only difference is the impact on `ObjectShorthandProps` interface
+
+### Option 1: Make `shorthands` optional
+
+This solves this problem by verifying if the shorthand is `undefined` or not.
+By verifying if shorthand is `undefined` we can opt for null rendering without compromising the cases which native slots don't have children
+
+- `getSlots` should stop verifying null rendering by the existence of the property `children`, instead depending on the existence of the shorthand itself
+
+```tsx
+// react-utilities/src/compose/getSlots.ts
+function getSlots(state, slotNames) {
+  // ...
+  if (state[name] === undefined) {
+    slots[name] = nullRender;
+    continue;
+  }
+  // ...
+}
+```
+
+- Since some slots may have default props, verification must be done on shorthand declaration to ensure that those slots will be null rendered even presenting default props
+
+```diff
+export function resolveShorthand(
+  value,
+  defaultProps,
++  {optional = true}
++) {
++  // verify if shorthand is undefined
++  if (value === null || (value === undefined && optional)) {
++    return undefined;
++  }
+}
+```
+
+The problem with this approach is that the shorthand signature now will carry with it an `undefined` value, making overrides a little less elegant:
+
+```ts
+export const CustomAccordionHeader = React.forwardRef<HTMLElement, AccordionHeaderProps>((props, ref) => {
+  const state = useAccordionHeader(props, ref);
+
+  if (state.icon) {
+    state.icon.onClick = () => {
+      /**
+       * this is an override,
+       * if this was a hook than it would need
+       * to be declared before the conditional
+       **/
+    };
+  }
+
+  useAccordionHeaderStyles(state);
+  return renderAccordionHeader(state);
+});
+```
+
+In cases where we simply want to ensure that the slot should be rendered even when `shorthand` is `undefined` than `{optional: false}` should be provided
+
+```ts
+export const useState = (props, ref) => {
+  return {
+    ...props,
+    slot1: resolveShorthand(props.slot1), // undefined
+    slot2: resolveShorthand(props.slot2, {}, { optional: false }), // {}
+  };
+};
+```
+
+### Option 2: Symbol for null rendering
```

Comment:
I like the idea of option 2, but why not use `null` as the indicator of a slot that shouldn't be rendered?

The two cases are:
1. **A slot that should only be rendered if the user provides content.** For example, Checkbox's `label` slot. In this case, the defaultProps argument to resolveShorthand could be null:
```ts
label: resolveShorthand(props.label, null),
```

If props.label is defined, that'd override the `null` default:
```jsx
<Checkbox label="Example" />
```

2. **A slot that should always be rendered by default, even if the user doesn't provide content.** In that case, the defaultProps argument would be non-null (an object with defaults, or an empty object `{}` if there are no additional defaults). For example, Checkbox's `input` slot:
```ts
input: resolveShorthand(props.input, { type: 'checkbox' }),
```

The user could override and prevent rendering the slot by passing null. This doesn't make sense in the case of Checkbox, but just as a demo:
```jsx
<Checkbox input={null} />
```

Reply:
That is a great alternative instead of a third argument! I'll add it to the RFC!

Reply:
That's case 2 in my example above. It provides default props `{ type: 'checkbox' }`, but can be explicitly set to `null` by the user: `input={null}`. To implement, `resolveShorthand` would just check the first argument for `null` and return `null` in that case.

There is a case 3, which is a slot that always must be rendered and can never be prevented from rendering (e.g. `root`). In that case, we'd simply not include `null` in the possible type for that slot.

```ts
alwaysRenderedSlot: React.HTMLAttributes<HTMLElement>;
optionallyRenderedSlot: React.HTMLAttributes<HTMLElement> | null;
```